lib/puppet/configurer.rb

```diff
@@ -247,132 +243,144 @@ def run(options = {})
+          # facts may be used to determine which catalog we get, we need to
+          # rerun the process if the environment is changed.
+          tries = 0
+          while catalog.environment and not catalog.environment.empty? and catalog.environment != @environment
+            if tries > 3
+              raise Puppet::Error, _("Catalog environment didn't stabilize after %{tries} fetches, aborting run") % { tries: tries }
+            end
+            Puppet.notice _("Local environment: '%{local_env}' doesn't match server specified environment '%{catalog_env}', restarting agent run with environment '%{catalog_env}'") % { local_env: @environment, catalog_env: catalog.environment }
+            @environment = catalog.environment
+            report.environment = @environment
 
-      # Here we set the local environment based on what we get from the
-      # catalog. Since a change in environment means a change in facts, and
-      # facts may be used to determine which catalog we get, we need to
-      # rerun the process if the environment is changed.
-      tries = 0
-      while catalog.environment and not catalog.environment.empty? and catalog.environment != @environment
-        if tries > 3
-          raise Puppet::Error, _("Catalog environment didn't stabilize after %{tries} fetches, aborting run") % { tries: tries }
-        end
-        Puppet.notice _("Local environment: '%{local_env}' doesn't match server specified environment '%{catalog_env}', restarting agent run with environment '%{catalog_env}'") % { local_env: @environment, catalog_env: catalog.environment }
-        @environment = catalog.environment
-        report.environment = @environment
+            query_options = get_facts(options)
+            query_options[:configured_environment] = configured_environment
 
-        query_options = get_facts(options)
-        query_options[:configured_environment] = configured_environment
+            return nil unless catalog = prepare_and_retrieve_catalog(options, query_options)
+            tries += 1
+          end
 
-        return nil unless catalog = prepare_and_retrieve_catalog(options, query_options)
-        tries += 1
-      end
+          if execute_prerun_command
+            prerun_command_successful = true
+            options[:report].code_id = catalog.code_id
+            options[:report].catalog_uuid = catalog.catalog_uuid
+            options[:report].cached_catalog_status = @cached_catalog_status
+            apply_catalog(catalog, options)
+          end
+        rescue => detail
+          Puppet.log_exception(detail, _("Failed to apply catalog: %{detail}") % { detail: detail })
+          return nil
+        end
+      end #thinmark
+    ensure
+      report.add_times(:total, run_internal_time)
+      report.finalize_report
 
-      execute_prerun_command or return nil
+      exit_status = report.exit_status if prerun_command_successful
+      exit_status = nil unless execute_postrun_command
```

Comment:
I'm having a little trouble following this logic. I know it's handling the logic pre

I *think* it might be a little easier to follow if it were written more like

```
if prerun_command_successful
  exit_status = report.exit status
  execute_postrun_command or return nil
  exit_status
end
```

but I'm not really sure the best way to change this. I just know when we look back at this later, it's going to take us a minute to parse out the what and why of this

Reply:
This is definitely a nitpick, and shouldn't block merging this in if we're worried about time

Reply:
@melissa So, the problem with that `if` form is that we always want `execute_postrun_command` to run, even if the prerun command failed, so we can't have execute_postrun_command in the body of that if. I'm not sure of a cleaner way of setting up the exit status where we have both the prerun, and postrun commands run regardless of other success/failure.